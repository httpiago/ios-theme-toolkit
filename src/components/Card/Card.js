import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * Card component.
 *
 * @since 0.2.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Card
 */
export default function Card({ children, shadow, outline, header, footer, className: aditionalClasses, ...rest }) {

  const prefix = 'card'
  const classes = mapClassList({
    [`${prefix}--shadow`]: shadow,
    [`${prefix}--outline`]: outline,
  })

  return (
    <div
      className={`${prefix} ${classes} ${aditionalClasses}`}
      {...rest}
    >
      {header && (
        <div className="card__header">{header}</div>
      )}
      <div className="card__content">{children}</div>
      {footer && (
        <div className="card__footer">{footer}</div>
      )}
    </div>
  )
}

Card.defaultProps = {
  className: '',
  outline: false,
  shadow: false,
  header: false,
  footer: false
}

Card.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** Enable long shadow */
  shadow: PropTypes.bool,
  /** Card without shadow */
  outline: PropTypes.bool,
  /** Anything that can be rendered by React */
  children: PropTypes.node,
  /** Card header (anything that can be rendered by React) */
  header: PropTypes.oneOfType([ PropTypes.node, PropTypes.bool ]),
  /** Card footer (anything that can be rendered by React) */
  footer: PropTypes.oneOfType([ PropTypes.node, PropTypes.bool ]),

}
