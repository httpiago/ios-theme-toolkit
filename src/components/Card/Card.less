@import '../../customizations.less';

// Default style
.card {
  background-color: #fff;
  -webkit-box-shadow: 0px 1px 2px rgba(0,0,0,.3);
  box-shadow: 0px 1px 2px rgba(0,0,0,.3);
  position: relative;
  border-radius: @base-border-radius;
  margin: @base-margin;
  font-family: @font-family;
  font-size: @font-size;
  line-height: 21px;
}

// Card sem sombra
.card--outline {
  -webkit-box-shadow: none;
  box-shadow: none;
  border: 1px solid rgba(0,0,0,.12);
}

// Card com sombra longa
.card--shadow {
  -webkit-box-shadow: 0 4px 16px rgba(0,0,0,.16);
  box-shadow: 0 4px 16px rgba(0,0,0,.16);
  border-radius: 8px;
  margin: 30px 20px;
}

.card__content {
  padding: @base-padding;
}

.card__header,
.card__footer {
  min-height: 44px;
  padding: (@base-padding - 5) @base-padding;
  position: relative;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -ms-flex-align: center;
  align-items: center;
}

.card__header {
  border-bottom: 1px solid #e1e1e1;
  border-radius: @base-border-radius @base-border-radius 0 0;
  font-size: 18px;
}

.card__footer {
  border-top: 1px solid #e1e1e1;
  border-radius: 0 0 @base-border-radius @base-border-radius;
  color: #6d6d72;
}
