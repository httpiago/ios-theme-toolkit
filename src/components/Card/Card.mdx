---
name: Card
route: /components/Card
menu: Components
---

import { Playground, PropsTable } from 'docz'
import { Card, Avatar, Button, ButtonGroup, Popover } from '../../index.js'
import '../../../dist/styles.css'

# Card

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { Card } from 'ios-theme-toolkit'

export default () => (
  <Card>Hello world!</Card>
)
```

## Examples

<Playground>
  <Card
    header={'Header'}
    footer={'Footer'}
    style={{ display: 'inline-block'}}
  >
    Default card style
  </Card>

  <Card outline style={{ display: 'inline-block'}}>Outline card</Card>

  <Card shadow style={{ display: 'inline-block'}}>Shadow card</Card>

  <Card style={{ maxWidth: 400 }}>
    <div style={{ display: 'flex', alignItems: 'center', marginBottom: 8 }}>
      <Avatar src="https://uinames.com/api/photos/male/6.jpg" round />
      <b style={{ marginLeft: 15, fontSize: 14, marginRight: 'auto' }}>James</b>

      <Popover closeOnInnerClick={true} content={(
        <ButtonGroup direction="vertical" outline>
          <Button>Copy post link</Button>
          <Button>Edit</Button>
          <Button>Delete</Button>
        </ButtonGroup>
      )}>
        <Button onlyIcon outline><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round"><circle cx="12" cy="12" r="1"/><circle cx="19" cy="12" r="1"/><circle cx="5" cy="12" r="1"/></svg></Button>
      </Popover>
    </div>

    <div>Hello world!</div>

    <img src="http://lorempixel.com/1000/600/nature/3/" style={{
      display: 'block',
      width: 'calc(100% + 30px)',
      margin: '8px -15px 15px'
    }}/>

    <div style={{ display: 'flex' }}>
      <Button color="red" outline>Like</Button>
      <Button color="green" outline>Comment</Button>
      <Button color="blue" outline>Share</Button>
    </div>
  </Card>
</Playground>

## Props

<PropsTable of={ Card } />

All the properties which are not listed above will be transferred to the \<div class="card"\> tag.
