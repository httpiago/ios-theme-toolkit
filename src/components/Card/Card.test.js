/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Card from './Card'

const COMPONENT = shallow(
  <Card
    header={<div className="check-header" />}
    footer={<div className="check-footer" />}
    className="aditionalClass"
  >
    Card content test
  </Card>
)

describe('Test Card component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ shadow: true })

    expect(COMPONENT.hasClass('card--shadow')).toBeTruthy()
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

  it('should render children correctly', () => {
    expect(COMPONENT.find('.card__content').text()).toEqual('Card content test')
  })
  
  it('should render header correctly', () => {
    expect(COMPONENT.exists('.check-header')).toBeTruthy()
  })
  
  it('should render footer correctly', () => {
    expect(COMPONENT.exists('.check-footer')).toBeTruthy()
  })

})
