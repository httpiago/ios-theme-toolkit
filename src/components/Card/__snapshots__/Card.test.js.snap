// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Card component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    className="aditionalClass"
    footer={
      <div
        className="check-footer"
      />
    }
    header={
      <div
        className="check-header"
      />
    }
    outline={false}
    shadow={false}
  >
    Card content test
  </Card>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card__header"
        >
          <div
            className="check-header"
          />
        </div>,
        <div
          className="card__content"
        >
          Card content test
        </div>,
        <div
          className="card__footer"
        >
          <div
            className="check-footer"
          />
        </div>,
      ],
      "className": "card  aditionalClass",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="check-header"
          />,
          "className": "card__header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "check-header",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Card content test",
          "className": "card__content",
        },
        "ref": null,
        "rendered": "Card content test",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="check-footer"
          />,
          "className": "card__footer",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "check-footer",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card__header"
          >
            <div
              className="check-header"
            />
          </div>,
          <div
            className="card__content"
          >
            Card content test
          </div>,
          <div
            className="card__footer"
          >
            <div
              className="check-footer"
            />
          </div>,
        ],
        "className": "card  aditionalClass",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="check-header"
            />,
            "className": "card__header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "check-header",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Card content test",
            "className": "card__content",
          },
          "ref": null,
          "rendered": "Card content test",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="check-footer"
            />,
            "className": "card__footer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "check-footer",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
